headerbp/signing.go

```diff
@@ -0,0 +1,179 @@
+package headerbp
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"slices"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/reddit/baseplate.go/secrets"
+	"github.com/reddit/baseplate.go/signing"
+)
+
+const delimiter = rune(0)
+
+var signatureVersionPrefix = strconv.Itoa(signatureVersion)
+
+var ErrInvalidSignatureVersion = fmt.Errorf("invalid  version")
+
+type headerSignatureContextKey struct{}
+
+func setHeaderSignature(ctx context.Context, sig string) context.Context {
+	return context.WithValue(ctx, headerSignatureContextKey{}, sig)
+}
+
+// HeaderSignatureFromContext gets the header signature from the context. This can be used in client middleware to propagate the
+// signature along with the headers if they are unchanged by the request.
+func HeaderSignatureFromContext(ctx context.Context) (string, bool) {
+	sig, ok := ctx.Value(headerSignatureContextKey{}).(string)
+	return sig, ok
+}
+
+var messageBufferPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
+func getBuffer() *bytes.Buffer {
+	buf := messageBufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	return buf
+}
+
+func putBuffer(buf *bytes.Buffer) {
+	messageBufferPool.Put(buf)
+}
+
+func concatHeaders(b *bytes.Buffer, headerNames []string, getHeader func(string) string, versionPrefix string) {
+	type manifestEntry struct {
+		// normalized is used to sort the manifest and as the header name in the message
+		normalized string
+
+		// original is maintained in order to potentially reduce the number of allocations when looking up the header
+		// value
+		original string
+	}
+	manifest := make([]manifestEntry, 0, len(headerNames))
+	for _, k := range headerNames {
+		manifest = append(manifest, manifestEntry{
+			// do not cache the normalized key here though since we do not know if we can trust it, and we don't want to
+			// cache junk headers.
+			normalized: normalizeKey(k, false),
+			original:   k,
+		})
+	}
+	slices.SortFunc(manifest, func(i, j manifestEntry) int {
+		if i.normalized < j.normalized {
+			return -1
+		} else if i.normalized > j.normalized {
+			return 1
+		} else {
+			return 0
+		}
```

Comment:
nit: this can be simplified into:
```suggestion
		return cmp.Compare(i.normalized, j.normalized)
```
(note: the stdlib cmp: https://pkg.go.dev/cmp)

also nit: for `slices` package usages I would use `left`/`right` over `i`/`j` as they are no longer indices, and `left`/`right` makes it much more obvious when you need to swap them for reverse order.